File: claude_rotator/rate_limit.py
"""Rate limit detection, reset time parsing, and per-account caching."""

from __future__ import annotations

import logging
import re
from datetime import datetime, timedelta, timezone

logger = logging.getLogger(__name__)

USAGE_LIMIT_PHRASES = [
    "out of extra usage",
    "usage limit",
    "rate limit",
]


class RateLimitCache:
    """Tracks per-account rate limit cooldowns with auto-expiry."""

    def __init__(self) -> None:
        self._until: dict[str | None, datetime] = {}

    def is_limited(self, account: str | None) -> bool:
        """Return True if the account is known to be rate-limited."""
        reset_time = self._until.get(account)
        if reset_time is None:
            return False
        if datetime.now(timezone.utc) >= reset_time:
            self._until.pop(account, None)
            return False
        return True

    def record(self, account: str | None, stdout: str, stderr: str) -> None:
        """Cache the rate limit reset time for an account."""
        reset_time = parse_reset_time(stdout, stderr)
        if reset_time:
            self._until[account] = reset_time
            label = account or "default"
            logger.info(f"Account '{label}' rate-limited until {reset_time.isoformat()}")
        else:
            fallback = datetime.now(timezone.utc) + timedelta(minutes=5)
            self._until[account] = fallback

    def clear(self, account: str | None = None) -> None:
        """Clear cached rate limit for one account, or all if account is not given."""
        if account is None:
            self._until.clear()
        else:
            self._until.pop(account, None)


def is_usage_limited(stdout: str, stderr: str) -> bool:
    """Check if the output indicates a usage limit error."""
    combined = (stdout + stderr).lower()
    return any(phrase in combined for phrase in USAGE_LIMIT_PHRASES)


def parse_reset_time(stdout: str, stderr: str) -> datetime | None:
    """Extract the rate limit reset time from the error message.

    Looks for patterns like "resets 8pm (UTC)" or "resets 2am (UTC)".
    Returns a UTC datetime, or None if unparseable.
    """
    combined = stdout + stderr
    match = re.search(r"resets\s+(\d{1,2})(am|pm)\s*\(UTC\)", combined, re.IGNORECASE)
    if not match:
        return None
    hour = int(match.group(1))
    ampm = match.group(2).lower()
    if ampm == "pm" and hour != 12:
        hour += 12
    elif ampm == "am" and hour == 12:
        hour = 0

    now = datetime.now(timezone.utc)
    reset = now.replace(hour=hour, minute=0, second=0, microsecond=0)
    if reset <= now:
        reset += timedelta(days=1)
    return reset

File: npm/src/rate-limit.ts
const USAGE_LIMIT_PHRASES = [
  "out of extra usage",
  "usage limit",
  "rate limit",
];

export function isUsageLimited(stdout: string, stderr: string): boolean {
  const text = stderr.toLowerCase();
  return USAGE_LIMIT_PHRASES.some((phrase) => text.includes(phrase));
}

export function parseResetTime(stdout: string, stderr: string): Date | null {
  const combined = stdout + stderr;
  const match = combined.match(/resets\s+(\d{1,2})(am|pm)\s*\(UTC\)/i);
  if (!match) return null;

  let hour = parseInt(match[1], 10);
  if (hour < 1 || hour > 12) return null;
  const ampm = match[2].toLowerCase();
  if (ampm === "pm" && hour !== 12) hour += 12;
  else if (ampm === "am" && hour === 12) hour = 0;

  const now = new Date();
  const reset = new Date(
    Date.UTC(
      now.getUTCFullYear(),
      now.getUTCMonth(),
      now.getUTCDate(),
      hour,
      0,
      0,
      0,
    ),
  );

  if (reset.getTime() <= now.getTime()) {
    reset.setUTCDate(reset.getUTCDate() + 1);
  }

  return reset;
}

export class RateLimitCache {
  private until = new Map<string, Date>();

  private keyFor(account: string | null): string {
    return account ?? "__default__";
  }

  isLimited(account: string | null): boolean {
    const key = this.keyFor(account);
    const resetTime = this.until.get(key);
    if (!resetTime) return false;
    if (Date.now() >= resetTime.getTime()) {
      this.until.delete(key);
      return false;
    }
    return true;
  }

  record(account: string | null, stdout: string, stderr: string): void {
    const key = this.keyFor(account);
    const resetTime = parseResetTime(stdout, stderr);
    if (resetTime) {
      this.until.set(key, resetTime);
    } else {
      const fallback = new Date(Date.now() + 5 * 60 * 1000);
      this.until.set(key, fallback);
    }
  }

  clear(account?: string | null): void {
    if (account === undefined) {
      this.until.clear();
    } else {
      this.until.delete(this.keyFor(account));
    }
  }
}
